Test/Sampler/test_sampler.py

```diff
@@ -0,0 +1,154 @@
+import netket as nk
+import networkx as nx
+import numpy as np
+import pytest
+from mpi4py import MPI
+from pytest import approx
+
+samplers = {}
+
+
+# TESTS FOR SPIN HILBERT
+# Constructing a 1d lattice
+g = nk.graph.Hypercube(length=6, ndim=1)
+
+# Hilbert space of spins from given graph
+hi = nk.hilbert.Spin(s=0.5, graph=g)
+ma = nk.machine.RbmSpin(hilbert=hi, alpha=1)
+ma.init_random_parameters(seed=1234, sigma=0.2)
+sa = nk.sampler.MetropolisLocal(machine=ma)
+
+samplers["MetropolisLocal RbmSpin"] = [sa, hi, ma]
+
+sa = nk.sampler.MetropolisLocalPt(machine=ma, n_replicas=4)
+samplers["MetropolisLocalPt RbmSpin"] = [sa, hi, ma]
+
+ha = nk.operator.Ising(hilbert=hi, h=1.0)
+sa = nk.sampler.MetropolisHamiltonian(machine=ma, hamiltonian=ha)
+samplers["MetropolisHamiltonian RbmSpin"] = [sa, hi, ma]
+
+ma = nk.machine.RbmSpinSymm(hilbert=hi, alpha=1)
+ma.init_random_parameters(seed=1234, sigma=0.2)
+sa = nk.sampler.MetropolisHamiltonianPt(
+    machine=ma, hamiltonian=ha, n_replicas=4)
+samplers["MetropolisHamiltonianPt RbmSpinSymm"] = [sa, hi, ma]
+
+hi = nk.hilbert.Boson(graph=g, n_max=4)
+ma = nk.machine.RbmSpin(hilbert=hi, alpha=1)
+ma.init_random_parameters(seed=1234, sigma=0.2)
+sa = nk.sampler.MetropolisLocal(machine=ma)
+g = nk.graph.Hypercube(length=4, ndim=1)
+samplers["MetropolisLocal Boson"] = [sa, hi, ma]
+
+sa = nk.sampler.MetropolisLocalPt(machine=ma, n_replicas=4)
+samplers["MetropolisLocalPt Boson"] = [sa, hi, ma]
+
+ma = nk.machine.RbmMultiVal(hilbert=hi, alpha=1)
+ma.init_random_parameters(seed=1234, sigma=0.2)
+sa = nk.sampler.MetropolisLocal(machine=ma)
+samplers["MetropolisLocal Boson MultiVal"] = [sa, hi, ma]
+
+hi = nk.hilbert.Spin(s=0.5, graph=g)
+g = nk.graph.Hypercube(length=6, ndim=1)
```

Comment:
@gcarleo you construct `hi` with the graph stored in `g` and now you assign a new graph to `g`. There are no more references to the old one and GC can thus collect it (which it does, or if not, you can force it to by doing `import gc; gc.collect()`). You then try to construct a machine which internally asks `hi` for the graph which is now a dangling reference -- and you get a segmentation fault.

P.S.: a word of advice, if I may, try compiling the C++ core with address sanitizer from Clang or GCC, they do a pretty good job at detecting such problems.

Reply:
Let's see if your theory is correct... :)

Reply:
Alright your theory is correct, chapeau @twesterhout ! 
Fine, but this reference situation is too error prone, the average user here would get crazy as much as I am getting crazy with these little problems... we need to find a decent solution to pass by value. 

And I agree for the sanitizer, that's why the new python setup file is much welcome ;)

Reply:
Maybe I'm seriously misreading it, but I though we agreed (at least for now) to follow @femtobit's suggestion and use `std::shared_ptr`s. I kind of assumed that's what you were working on @gcarleo...

Reply:
But would that solve the problem entirely? That's the question

Reply:
In the layers case, I am passing a vector of shared pointers to the FFNN class, and I still had a seg fault, if you remember

Reply:
> But would that solve the problem entirely? That's the question

Yeah, it really should, if done correctly.

> In the layers case, I am passing a vector of shared pointers to the FFNN class, and I still had a seg fault, if you remember

I haven't looked at the code carefully, but I really hope it falls under the "not (yet) implemented correctly" clause :)